File: arbol_grafos/src/main/java/dao/ArbolAVL.java
package dao;

import dto.AVL_Nodo;

public class ArbolAVL {
    public AVL_Nodo raiz;
    
    public void ArbolAVL(){
        raiz = null;
    }
    
    public int altura(AVL_Nodo puntero){
        if(puntero == null){
            return 0;
        } else {
            return puntero.getAltura();
        }
    }
    
    public int FE(AVL_Nodo puntero){
        return altura(puntero.getDerecha()) - altura(puntero.getIzquierda());
    }
    
    public AVL_Nodo giroIzquierda(AVL_Nodo puntero){
        AVL_Nodo aux = puntero.getDerecha();
        puntero.setDerecha(aux.getIzquierda());
        aux.setIzquierda(puntero);
        
        puntero.setAltura(Math.max(altura(puntero.getIzquierda()),altura(puntero.getIzquierda()))+1);
        aux.setAltura(Math.max(altura(aux.getIzquierda()), altura(aux.getIzquierda()))+1);
        return aux;
    }
    
    public AVL_Nodo giroDerecha(AVL_Nodo puntero){
        AVL_Nodo aux = puntero.getIzquierda();
        puntero.setIzquierda(aux.getDerecha());
        aux.setDerecha(puntero);
        
        puntero.setAltura(Math.max(altura(puntero.getIzquierda()),altura(puntero.getIzquierda()))+1);
        aux.setAltura(Math.max(altura(aux.getIzquierda()), altura(aux.getIzquierda()))+1);
        return aux;
    }
    
    public AVL_Nodo equilibrar(AVL_Nodo puntero){
        if (FE(puntero) > 1) {
            if (FE(puntero.getDerecha()) < 0) {
                puntero.setDerecha(giroDerecha(puntero.getDerecha()));
            }
            puntero = giroIzquierda(puntero);
        } else if (FE(puntero) < -1) {
            if (FE(puntero.getIzquierda()) > 0) {
                puntero.setIzquierda(giroIzquierda(puntero.getIzquierda()));
            }
            puntero = giroDerecha(puntero);
        }
        return puntero;
    }
    
    public AVL_Nodo agregar(AVL_Nodo puntero, int valor){
        if(puntero == null){
            return new AVL_Nodo(valor);
        } else {
            if(valor < puntero.getValor()){
                puntero.setIzquierda(agregar(puntero.getIzquierda(),valor));
            } else if (valor > puntero.getValor()){
                puntero.setDerecha(agregar(puntero.getDerecha(),valor));
            } else {
                puntero.incrementarContador();
            }
        }
        puntero.setAltura(Math.max(altura(puntero.getIzquierda()),altura(puntero.getDerecha())) + 1);
        return equilibrar(puntero);
    }
    
    public void preorden(AVL_Nodo puntero, int nivel){
        if(puntero != null){
            System.out.println(puntero.getValor() + " - Altura: " + puntero.getAltura() + " - Nivel: " + nivel + " - Contador: " + puntero.getContador());
            preorden(puntero.getIzquierda(),nivel+1);
            preorden(puntero.getDerecha(),nivel+1);
        }
    }
    
    public void inorden(AVL_Nodo puntero, int nivel) {
        if (puntero != null) {
            inorden(puntero.getIzquierda(), nivel + 1);
            System.out.println(puntero.getValor() + " - Altura: " + puntero.getAltura() + " - Nivel: " + nivel + " - Contador: " + puntero.getContador());
            inorden(puntero.getDerecha(), nivel + 1);
        }
    }

    
    public void posorden(AVL_Nodo puntero, int nivel) {
        if (puntero != null) {
            posorden(puntero.getIzquierda(), nivel + 1);
            posorden(puntero.getDerecha(), nivel + 1);
            System.out.println(puntero.getValor() + " - Altura: " + puntero.getAltura() + " - Nivel: " + nivel + " - Contador: " + puntero.getContador());
        }
    }
}

File: arbol_grafos_web/src/main/webapp/arbol-avl.js

    class Nodo {
      constructor(valor){
        this.valor = valor;
        this.izq = null;
        this.der = null;
        this.altura = 1;
        this.contador = 1;
      }
    }

    class ArbolAVL {
      constructor() {
        this.raiz = null;
      }

      altura(nodo) {
        return nodo ? nodo.altura : 0;
      }

      FE(nodo) {
        return this.altura(nodo.der) - this.altura(nodo.izq);
      }

      giroIzquierda(nodo) {
        const aux = nodo.der;
        nodo.der = aux.izq;
        aux.izq = nodo;

        nodo.altura = 1 + Math.max(this.altura(nodo.izq), this.altura(nodo.der));
        aux.altura = 1 + Math.max(this.altura(aux.izq), this.altura(aux.der));

        return aux;
      }

      giroDerecha(nodo) {
        const aux = nodo.izq;
        nodo.izq = aux.der;
        aux.der = nodo;

        nodo.altura = 1 + Math.max(this.altura(nodo.izq), this.altura(nodo.der));
        aux.altura = 1 + Math.max(this.altura(aux.izq), this.altura(aux.der));

        return aux;
      }

      equilibrar(nodo) {
        if (this.FE(nodo) > 1) {
          if (this.FE(nodo.der) < 0) {
            nodo.der = this.giroDerecha(nodo.der);
          }
          nodo = this.giroIzquierda(nodo);
        } else if (this.FE(nodo) < -1) {
          if (this.FE(nodo.izq) > 0) {
            nodo.izq = this.giroIzquierda(nodo.izq);
          }
          nodo = this.giroDerecha(nodo);
        }
        return nodo;
      }

      agregar(nodo, valor) {
        if (!nodo) return new Nodo(valor);

        if (valor < nodo.valor) {
          nodo.izq = this.agregar(nodo.izq, valor);
        } else if (valor > nodo.valor) {
          nodo.der = this.agregar(nodo.der, valor);
        } else {
          nodo.contador++;
        }

        nodo.altura = 1 + Math.max(this.altura(nodo.izq), this.altura(nodo.der));
        return this.equilibrar(nodo);
      }

//      preorden(nodo, nivel = 0, salida = []) {
//        if (nodo) {
//          salida.push(`${'  '.repeat(nivel)}${nodo.valor} (Altura: ${nodo.altura}, Contador: ${nodo.contador})`);
//          this.preorden(nodo.izq, nivel + 1, salida);
//          this.preorden(nodo.der, nivel + 1, salida);
//        }
//        return salida;
//      }

inorden(nodo, nivel = 0, salida = []) {
  if (nodo) {
    this.inorden(nodo.izq, nivel + 1, salida);  // 1. Izquierda
    salida.push(`${'  '.repeat(nivel)}${nodo.valor} (Altura: ${nodo.altura}, Contador: ${nodo.contador})`);  // 2. Actual
    this.inorden(nodo.der, nivel + 1, salida);  // 3. Derecha
  }
  return salida;
}



    }

    const arbol = new ArbolAVL();

    function insertar() {
      const valor = parseInt(document.getElementById('valor').value);
      if (!isNaN(valor)) {
        arbol.raiz = arbol.agregar(arbol.raiz, valor);
//        const resultado = arbol.preorden(arbol.raiz).join('\n');
//        document.getElementById('salida').textContent = resultado;
const resultado = arbol.inorden(arbol.raiz).join('\n');
document.getElementById('salida').textContent = resultado;
      }
    }

